packages/hulypulse-client/src/client.ts

```diff
@@ -0,0 +1,383 @@
+  private closed = false
+  private reconnectTimeout: any | undefined
+  private readonly RECONNECT_INTERVAL_MS = 1000
+  private subscribes: Record<string, Callback[]> = {}
+
+  private pingTimeout: ReturnType<typeof setTimeout> | undefined
+  private pingInterval: ReturnType<typeof setInterval> | undefined
+  private readonly PING_INTERVAL_MS = 30 * 1000
+  private readonly PING_TIMEOUT_MS = 5 * 60 * 1000
+  private readonly SEND_TIMEOUT_MS = 3000
+
+  private correlationId = 1
+
+  private pending = new Map<
+    number,
+    {
+      resolve: (v: any) => void
+      reject: (e: any) => void
+      send_timeout: ReturnType<typeof setTimeout>
+    }
+  >()
+
+  private constructor(private readonly url: string | URL) {}
+
+  private async connect(): Promise<void> {
+    return new Promise((resolve, reject) => {
+      const ws = new WebSocket(this.url.toString())
+      this.ws = ws
+
+      ws.onopen = () => {
+        this.resubscribe(reject)
+        this.startPing()
+        resolve()
+      }
+
+      ws.onerror = (event) => {
+        this.reconnect()
+      }
+
+      ws.onclose = (event) => {
+        this.reconnect()
+      }
+
+      ws.onmessage = (event) => {
+        try {
+          if (event.data === WEBSOCKET_MESSAGE_CONSTANT_PING_REQUEST) {
+            this.ws?.send(WEBSOCKET_MESSAGE_CONSTANT_PONG_RESPONSE)
+            return
+          }
+
+          if (event.data === WEBSOCKET_MESSAGE_CONSTANT_PONG_RESPONSE) {
+            clearTimeout(this.pingTimeout)
+            return
+          }
+
+          const msg = JSON.parse(event.data.toString())
+          // Handle incoming messages (Set, Expired, Del)
+          if (msg.message) {
+            for (const key in this.subscribes) {
+              if (msg.key.startsWith(key)) {
+                this.subscribes[key].forEach((cb, index) => {
+                  try {
+                    cb(msg.key, msg.message === 'Set' ? msg.value : undefined, {
+                      msg: msg,
+                      subscribed_key: key,
+                      run_index: index
+                    })
+                  } catch (err) {
+                    console.error(`Error in callback #${index} with key "${key}":`, err)
+                  }
+                })
+              }
+            }
+          }
+          // Handle correlation responses
+          else if (msg.correlation) {
+            let id = Number(msg.correlation)
+            if (id && this.pending.has(id)) {
+              clearTimeout(this.pending.get(id)!.send_timeout)
+              this.pending.get(id)!.resolve(msg)
+              this.pending.delete(id)
+            }
+          }
+          // Handle unknown
+          else {
+            console.warn('Unknown message format:', msg)
+          }
+        } catch (e) {
+          console.error('Failed to parse message', e)
+        }
+      }
+    })
+  }
+
+  private async resubscribe(reject: (reason?: any) => void): Promise<void> {
```

Comment:
reject is unused?

1. The function seem to be not asynchronous, but is marked as `async`
2. It uses asynchronous functions, but does not properly wait for them (or mark usages with `void`)

This will cause a problem with unhandled rejections:
1. `resubscribe` calls `send` and immediately exits
2. `send` throws an error which is catched, but new error is thrown
3. this error won't be handled